server/src/main/java/io/deephaven/server/appmode/ApplicationServiceGrpcImpl.java

```diff
@@ -67,149 +67,264 @@ public ApplicationServiceGrpcImpl(final AppMode mode,
-                updatedFields.add(id);
-            }
+        for (Entry<String, String> e : changes.updated.entrySet()) {
+            final String key = e.getKey();
+            final String type = e.getValue();
+            final AppFieldId id = AppFieldId.fromScopeName(key);
+            update(id, QUERY_SCOPE_DESCRIPTION, type);
         }
 
-        for (final String name : changes.created.keySet()) {
-            final AppFieldId id = AppFieldId.fromScopeName(name);
-            final Object value = scriptSession.unwrapObject(scriptSession.getVariable(name));
-            final ScopeField field = new ScopeField(name, value);
-            final FieldInfo fieldInfo = getFieldInfo(id, field);
-            if (fieldInfo == null) {
-                // The script session should not have told us about this variable...
-                throw new IllegalStateException(
-                        String.format("Field information could not be generated for scope variable '%s'", name));
-            }
-            final Field<?> oldField = addedFields.put(id, field);
-            if (oldField != null) {
-                throw new IllegalStateException(
-                        String.format("Script session notified of new field but was already existing '%s'", name));
-            }
+        for (Entry<String, String> e : changes.created.entrySet()) {
+            final String key = e.getKey();
+            final String type = e.getValue();
+            final AppFieldId id = AppFieldId.fromScopeName(key);
+            create(id, QUERY_SCOPE_DESCRIPTION, type);
         }
 
         schedulePropagationOrClearIncrementalState();
     }
 
+    // todo: better interface to sync across large change
+
     @Override
-    public synchronized void onRemoveField(ApplicationState app, Field<?> oldField) {
+    public synchronized void onNewField(final ApplicationState app, final Field<?> field) {
         if (!mode.hasVisibilityToAppExports()) {
             return;
         }
-
-        final AppFieldId id = AppFieldId.from(app, oldField.name());
-        Field<?> recentlyAdded = addedFields.remove(id);
-        if (recentlyAdded != null) {
-            tracker.maybeUnmanage(recentlyAdded.value());
-            return;
-        }
-        updatedFields.remove(id);
-        removedFields.add(id);
-
+        final AppFieldId id = AppFieldId.from(app, field.name());
+        final String type = typeLookup.type(field.value()).orElse(null);
+        create(id, field.description().orElse(null), type);
         schedulePropagationOrClearIncrementalState();
     }
 
     @Override
-    public synchronized void onNewField(final ApplicationState app, final Field<?> field) {
+    public synchronized void onRemoveField(ApplicationState app, Field<?> oldField) {
         if (!mode.hasVisibilityToAppExports()) {
             return;
         }
+        final AppFieldId id = AppFieldId.from(app, oldField.name());
+        // todo: or should we rely on the known type
```

Comment:
I think that is "safest". Same for description. Might be annoying given the current code pattern, though.

Reply:
This is actually a bug to not use the known type. Updated code is much cleaner.